README.md

<img src="https://raw.githubusercontent.com/reconhub/incidence/master/artwork/banner.png">

<br>

[![Travis-CI Build
Status](https://travis-ci.org/reconhub/incidence.svg?branch=master)](https://travis-ci.org/reconhub/incidence)
[![Build
status](https://ci.appveyor.com/api/projects/status/7h2mgej230dv5r7w/branch/master?svg=true)](https://ci.appveyor.com/project/thibautjombart/incidence/branch/master)
[![Coverage
Status](https://codecov.io/github/reconhub/incidence/coverage.svg?branch=master)](https://codecov.io/github/reconhub/incidence?branch=master)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/incidence)](https://cran.r-project.org/package=incidence)
[![CRAN
Downloads](https://cranlogs.r-pkg.org/badges/incidence)](https://cran.r-project.org/package=incidence)
[![Downloads from Rstudio
mirror](https://cranlogs.r-pkg.org/badges/grand-total/incidence)](http://www.r-pkg.org/pkg/incidence)

[![DOI](https://zenodo.org/badge/49890366.svg)](https://zenodo.org/badge/latestdoi/49890366)

# Installing the package

To install the current stable, CRAN version of the package, type:

``` r
install.packages("incidence")
```

To benefit from the latest features and bug fixes, install the
development, *github* version of the package using:

``` r
devtools::install_github("reconhub/incidence")
```

Note that this requires the package *devtools* installed.

# What does it do?

The main features of the package include:

  - **`incidence`**: compute incidence from dates in various formats;
    any fixed time interval can be used; the returned object is an
    instance of the (S3) class *incidence*.

  - **`plot`**: this method (see `?plot.incidence` for details) plots
    *incidence* objects, and can also add predictions of the model(s)
    contained in an *incidence\_fit* object (or a list of such objects).

  - **`fit`**: fit one or two exponential models (i.e. linear regression
    on log-incidence) to an *incidence* object; two models are
    calibrated only if a date is provided to split the time series in
    two (argument `split`); this is typically useful to model the two
    phases of exponential growth, and decrease of an outbreak; each
    model returned is an instance of the (S3) class *incidence\_fit*,
    each of which contains various useful information (e.g. growth rate
    *r*, doubling/halving time, predictions and confidence intervals);
    results can be plotted using `plot`, or added to an existing
    `uncudence` plot using the piping-friendly function
    `add_incidence_fit`.

  - **`fit_optim_split`**: finds the optimal date to split the time
    series in two, typically around the peak of the epidemic.

  - **`[`**: lower-level subsetan of *incidence* objects, permiting to
    specify which dates and groups to retain; uses a syntax similar to
    matrices, i.e. `x[i, j]`, where `x` is the *incidence* object, `i` a
    subset of dates, and `j` a subset of groups.

  - **`subset`**: subset an *incidence* object by specifying a time
    window.

  - **`pool`**: pool incidence from different groups into one global
    incidence time series.

  - **`cumulate`**: computes cumulative incidence over time from and
    `incidence` object.

  - **`as.data.frame`**: converts an *incidence* object into a
    `data.frame` containing dates and incidence values.

  - **`bootstrap`**: generates a bootstrapped *incidence* object by
    re-sampling, with replacement, the original dates of events.

  - **`find_peak`**: locates the peak time of the epicurve.

  - **`estimate_peak`**: uses bootstrap to estimate the peak time (and
    related confidence interval) of a partially observed outbreak.

# Resources

## Vignettes

An overview of *incidence* is provided below in the worked example
below. More detailed tutorials are distributed as vignettes with the
package:

``` r
vignette("overview", package="incidence")
vignette("customize_plot", package="incidence")
vignette("incidence_class", package="incidence")
vignette("incidence_fit_class", package="incidence")
vignette("conversions", package="incidence")
```

## Websites

The following websites are available:

  - The official *incidence* website, providing an overview of the
    package’s functionalities, up-to-date tutorials and documentation:
    <br> <https://www.repidemicsconsortium.org/incidence>

  - The *incidence* project on *github*, useful for developers,
    contributors, and users wanting to post issues, bug reports and
    feature requests: <br> <https://github.com/reconhub/incidence>

  - The *incidence* page on CRAN: <br>
    <https://CRAN.R-project.org/package=incidence>

## Getting help online

Bug reports and feature requests should be posted on *github* using the
[*issue* system](https://github.com/reconhub/incidence/issues). All
other questions should be posted on the **RECON forum**: <br>
<https://www.repidemicsconsortium.org/forum/>

# A quick overview

The following worked example provides a brief overview of the package’s
functionalities. See the [*vignettes section*](#vignettes) for more
detailed tutorials.

## Loading the data

This example uses the simulated Ebola Virus Disease (EVD) outbreak from
the package [*outbreaks*](https://github.com/reconhub/outbreaks). We
will compute incidence for various time steps, calibrate two exponential
models around the peak of the epidemic, and analyse the results.

First, we load the data:

``` r
library(outbreaks)
library(ggplot2)
library(incidence)

dat <- ebola_sim$linelist$date_of_onset
class(dat)
#> [1] "Date"
head(dat)
#> [1] "2014-04-07" "2014-04-15" "2014-04-21" "2014-04-27" "2014-04-26"
#> [6] "2014-04-25"
```

## Computing and plotting incidence

We compute the weekly incidence:

``` r
i.7 <- incidence(dat, interval = 7)
i.7
#> <incidence object>
#> [5888 cases from days 2014-04-07 to 2015-04-27]
#> [5888 cases from ISO weeks 2014-W15 to 2015-W18]
#> 
#> $counts: matrix with 56 rows and 1 columns
#> $n: 5888 cases in total
#> $dates: 56 dates marking the left-side of bins
#> $interval: 7 days
#> $timespan: 386 days
#> $cumulative: FALSE
plot(i.7)
```

![](figs/incid1-1.png)<!-- -->

`incidence` can also compute incidence by specified groups using the
`groups` argument. For instance, we can compute the weekly incidence by
gender:

``` r
i.7.sex <- incidence(dat, interval = "week", groups = ebola_sim$linelist$gender)
i.7.sex
#> <incidence object>
#> [5888 cases from days 2014-04-07 to 2015-04-27]
#> [5888 cases from ISO weeks 2014-W15 to 2015-W18]
#> [2 groups: f, m]
#> 
#> $counts: matrix with 56 rows and 2 columns
#> $n: 5888 cases in total
#> $dates: 56 dates marking the left-side of bins
#> $interval: 1 week
#> $timespan: 386 days
#> $cumulative: FALSE
plot(i.7.sex, stack = TRUE, border = "grey")
```

![](figs/gender-1.png)<!-- -->

## Handling `incidence` objects

`incidence` objects can be manipulated easily. The `[` operator
implements subetting of dates (first argument) and groups (second
argument). For instance, to keep only the first 20 weeks of the
epidemic:

``` r
i.7[1:20]
#> <incidence object>
#> [797 cases from days 2014-04-07 to 2014-08-18]
#> [797 cases from ISO weeks 2014-W15 to 2014-W34]
#> 
#> $counts: matrix with 20 rows and 1 columns
#> $n: 797 cases in total
#> $dates: 20 dates marking the left-side of bins
#> $interval: 7 days
#> $timespan: 134 days
#> $cumulative: FALSE
plot(i.7[1:20])
```

![](figs/start-1.png)<!-- -->

Some temporal subsetting can be even simpler using `subset`, which
permits to retain data within a specified time window:

``` r
i.tail <- subset(i.7, from = as.Date("2015-01-01"))
i.tail
#> <incidence object>
#> [1156 cases from days 2015-01-05 to 2015-04-27]
#> [1156 cases from ISO weeks 2015-W02 to 2015-W18]
#> 
#> $counts: matrix with 17 rows and 1 columns
#> $n: 1156 cases in total
#> $dates: 17 dates marking the left-side of bins
#> $interval: 7 days
#> $timespan: 113 days
#> $cumulative: FALSE
plot(i.tail, border = "white")
```

![](figs/tail-1.png)<!-- -->

Subsetting groups can also matter. For instance, let’s try and visualise
the incidence based on onset of symptoms by outcome:

``` r
i.7.outcome <- incidence(dat, "week", groups = ebola_sim$linelist$outcome)
i.7.outcome
#> <incidence object>
#> [5888 cases from days 2014-04-07 to 2015-04-27]
#> [5888 cases from ISO weeks 2014-W15 to 2015-W18]
#> [3 groups: Death, Recover, NA]
#> 
#> $counts: matrix with 56 rows and 3 columns
#> $n: 5888 cases in total
#> $dates: 56 dates marking the left-side of bins
#> $interval: 1 week
#> $timespan: 386 days
#> $cumulative: FALSE
plot(i.7.outcome, stack = TRUE, border = "grey")
```

![](figs/i7outcome-1.png)<!-- -->

To visualise the cumulative incidence:

``` r
i.7.outcome.c <- cumulate(i.7.outcome)
i.7.outcome.c
#> <incidence object>
#> [5888 cases from days 2014-04-07 to 2015-04-27]
#> [5888 cases from ISO weeks 2014-W15 to 2015-W18]
#> [3 groups: Death, Recover, NA]
#> 
#> $counts: matrix with 56 rows and 3 columns
#> $n: 5888 cases in total
#> $dates: 56 dates marking the left-side of bins
#> $interval: 1 week
#> $timespan: 386 days
#> $cumulative: TRUE
plot(i.7.outcome.c)
```

![](figs/i7outcome_cum-1.png)<!-- -->

Groups can also be collapsed into a single time series using `pool`:

``` r
i.pooled <- pool(i.7.outcome)
i.pooled
#> <incidence object>
#> [5888 cases from days 2014-04-07 to 2015-04-27]
#> [5888 cases from ISO weeks 2014-W15 to 2015-W18]
#> 
#> $counts: matrix with 56 rows and 1 columns
#> $n: 5888 cases in total
#> $dates: 56 dates marking the left-side of bins
#> $interval: 1 week
#> $timespan: 386 days
#> $cumulative: FALSE
identical(i.7$counts, i.pooled$counts)
#> [1] TRUE
```

## Modelling incidence

Incidence data, excluding zeros, can be modelled using log-linear
regression of the form: log(*y*) = *r* x *t* + *b*

where *y* is the incidence, *r* is the growth rate, *t* is the number of
days since a specific point in time (typically the start of the
outbreak), and *b* is the intercept.

Such model can be fitted to any incidence object using `fit`. Of course,
a single log-linear model is not sufficient for modelling our time
series, as there is clearly an growing and a decreasing phase. As a
start, we can calibrate a model on the first 20 weeks of the epidemic:

``` r
plot(i.7[1:20])
```

![](figs/fit1-1.png)<!-- -->

``` r
early.fit <- fit(i.7[1:20])
early.fit
#> <incidence_fit object>
#> 
#> $model: regression of log-incidence over time
#> 
#> $info: list containing the following items:
#>   $r (daily growth rate):
#> [1] 0.03175771
#> 
#>   $r.conf (confidence interval):
#>           2.5 %     97.5 %
#> [1,] 0.02596229 0.03755314
#> 
#>   $doubling (doubling time in days):
#> [1] 21.8261
#> 
#>   $doubling.conf (confidence interval):
#>         2.5 %   97.5 %
#> [1,] 18.45777 26.69823
#> 
#>   $pred: data.frame of incidence predictions (20 rows, 5 columns)
```

The resulting objects can be plotted, in which case the prediction and
its confidence interval is displayed:

``` r
plot(early.fit)
```

![](figs/fit-plot-1-1.png)<!-- -->

However, a better way to display these predictions is adding them to the
incidence plot using the argument `fit`:

``` r
plot(i.7[1:20], fit = early.fit)
```

![](figs/fit-plot-2-1.png)<!-- -->

Alternatively, these can be piped using:

``` r
library(magrittr)
plot(i.7[1:20]) %>% add_incidence_fit(early.fit)
```

![](figs/pipe-fit-1.png)<!-- -->

In this case, we would ideally like to fit two models, before and after
the peak of the epidemic. This is possible using the following approach,
in which the best possible splitting date (i.e. the one maximizing the
average fit of both models), is determined automatically:

``` r
best.fit <- fit_optim_split(i.7)
best.fit
#> $df
#>         dates   mean.R2
#> 1  2014-08-04 0.7650406
#> 2  2014-08-11 0.8203351
#> 3  2014-08-18 0.8598316
#> 4  2014-08-25 0.8882682
#> 5  2014-09-01 0.9120857
#> 6  2014-09-08 0.9246023
#> 7  2014-09-15 0.9338797
#> 8  2014-09-22 0.9339813
#> 9  2014-09-29 0.9333246
#> 10 2014-10-06 0.9291131
#> 11 2014-10-13 0.9232523
#> 12 2014-10-20 0.9160439
#> 13 2014-10-27 0.9071665
#> 
#> $split
#> [1] "2014-09-22"
#> 
#> $fit
#> <list of incidence_fit objects>
#> 
#> attr(x, 'locations'): list of vectors with the locations of each incidence_fit object
#> 
#> 'before'
#> 'after'
#> 
#> $model: regression of log-incidence over time
#> 
#> $info: list containing the following items:
#>   $r (daily growth rate):
#>      before       after 
#>  0.02982209 -0.01016191 
#> 
#>   $r.conf (confidence interval):
#>              2.5 %       97.5 %
#> before  0.02608945  0.033554736
#> after  -0.01102526 -0.009298561
#> 
#>   $doubling (doubling time in days):
#>   before 
#> 23.24274 
#> 
#>   $doubling.conf (confidence interval):
#>           2.5 %  97.5 %
#> before 20.65721 26.5681
#> 
#>   $halving (halving time in days):
#>    after 
#> 68.21031 
#> 
#>   $halving.conf (confidence interval):
#>          2.5 %   97.5 %
#> after 62.86899 74.54349
#> 
#>   $pred: data.frame of incidence predictions (57 rows, 6 columns)
#> 
#> $plot
```

![](figs/optim-1.png)<!-- -->

``` r
plot(i.7, fit = best.fit$fit)
```

![](figs/optim-2.png)<!-- -->

# Credits

  - Contributors (by alphabetic order):
  - [Sangeeta Bhatia](https://github.com/sangeetabhatia03)
  - [Jun Cai](https://github.com/caijun)
  - [Rich Fitzjohn](https://github.com/richfitz)
  - [Thibaut Jombart](https://github.com/thibautjombart)
  - [Zhian Kamvar](https://github.com/zkamvar)
  - [Juliet RC Pulliam](https://github.com/jrcpulliam)
  - [Jakob Schumacher](https://github.com/jakobschumacher)

See details of contributions on: <br>
<https://github.com/reconhub/incidence/graphs/contributors>

Contributions are welcome via **pull requests**.

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.

**Maintainer:** Zhian N. Kamvar (<zkamvar@gmail.com>)


NEWS.md
incidence 1.7.3
============================

* change of maintainer only.

incidence 1.7.2
============================

### BUG FIXES

* `plot.incidence()` now reverts to the previous behaviour of plotting ticks on 
  either side of the interval as opposed to centering within the interval. 
* The default version for {ggplot2} has been set to 3.3.2. 

incidence 1.7.1
============================

### BUG FIXES

* Fix for a [bug in](https://github.com/reconhub/incidence/issues/119) `plot.incidence()` 
introduced with new release of *ggplot2* ([bug report](https://github.com/tidyverse/ggplot2/issues/3873))
As a temporary fix, dates are now centered within the interval instead of to the left of the interval. 



incidence 1.7.0 
============================

### NEW FEATURES

 * Any interval `seq.Date()` can handle (e.g. "5 weeks") can be handled by
   `incidence()` (see https://github.com/reconhub/incidence/issues/67)
 * Weekly intervals can start on any day of the week by allowing things like 
   "epiweek", "isoweek", "wednesday week", "2 Saturday weeks", etc.
   (see https://github.com/reconhub/incidence/issues/55#issuecomment-405297526)
 * the item `$weeks` is now added to the incidence object, which contains an
   "aweek" class
 * plotting will now force the first tick to be the starting point of the
   incidence curve

### NEW FUNCTIONS

 * `make_breaks()` will automatically calculate breaks from an incidence object
   for plotting. 
 * `scale_x_incidence()` will produce a ggplot2 "ScaleContinuous" object to add
   to a ggplot.

### DEPRECATED

 * `plot.incidence()` argument `labels_iso` is deprecated in favor of 
   `labels_week`
 * Incidence objects will still have `$isoweeks` if the weeks are ISO 8601
   standard, but users should rely intead on `$weeks` instead. The `$isoweeks`
   element will be removed in a future version of incidence.
 * `as.incidence()` argument `isoweeks` has been deprecated in favour of 
   `standard`

### DEPENDENCIES

 - ISOweek import changed to [aweek](https://www.repidemicsconsortium.org/aweek/)

### Documentation

 - Vignettes have been updated with examples.

incidence 1.6.0 (2019-03-05) 
============================

### BEHAVIORAL CHANGE

* `incidence()` will no longer allow a non-standard `first_date` to override
  `standard = TRUE. The first call to `incidence()` specifying `first_date` 
  without `standard` will issue a warning. To use non-standard first dates, 
  specify `standard = FALSE`. To remove the warning, use
  `options(incidence.warn.first_date = FALSE)`. See
  https://github.com/reconhub/incidence/issues/87 for details.  

### MISC

* `citation("incidence")` will now give the proper citation for our article in
  F1000 research and the global DOI for archived code. See
  https://github.com/reconhub/incidence/pulls/106
* Tests have been updated to avoid randomisation errors on R 3.6.0
  See https://github.com/reconhub/incidence/issues/107


incidence 1.5.4 (2019-01-15)
============================

### BUG FIX

* `incidence()` now returns an error when supplied a character vector that is
  not formatted as (yyyy-mm-dd).
  (See https://github.com/reconhub/incidence/issues/88)
* `fit()` now returns correct coefficients when dates is POSIXt by converting to
  Date. (See https://github.com/reconhub/incidence/issues/91)
* `plot.incidence()` now plots in UTC by default for POSIXt incidence objects.
  this prevents a bug where different time zones would cause a shift in the bars
  (See https://github.com/reconhub/incidence/issues/99).

### MISC

* A test that randomly failed on CRAN has been fixed.
  (See https://github.com/reconhub/incidence/issues/95).
* Plotting tests have been updated for new version of vdiffr
  (See https://github.com/reconhub/incidence/issues/96).
* POSIXct incidence are first passed through POSIXlt when initialized.
* A more informative error message is generated for non ISO 8601 formatted 
  `first_date` and `last_date` parameters.

incidence 1.5.3 (2018-12-07)
============================

### BUG FIX

* `plot.incidence()` will now respect single groups.
  (See https://github.com/reconhub/incidence/issues/84)
* `as.data.frame.incidence()` will now respect single groups.
  (See https://github.com/reconhub/incidence/issues/84)

### MISC

* `demo("incidence-demo" package = "incidence")` has been updated to show use of
  custom colors.

incidence 1.5.2 (2018-11-30)
============================

### BUG FIX

* `print.incidence()` will now print isoweeks even if the `$interval` element is
  "week".

### MISC

* `subset.incidence()` will now give a more informative error message when the
  user specifies a group that does not exist. 
* `demo('incidence-demo', package = 'incidence')` now shows plotting with 
  `show_cases = TRUE`.
* In the the case where a date is accidentally mis-typed leading to a gross
  mis-calculation of the date range (i.e. 2018 is mis-typed as 3018), a warning
  will be issued. The default threshold is set at 18262 days (50 years), but the
  user can define their own threshold by setting the `incidence.max.days` option

incidence 1.5.1 (2018-11-14)
============================

### BUG FIX

* Two bugs regarding the ordering of groups when the user specifies a factor/
  column order have been fixed. This affects `plot.incidence()`, `incidence()`,
  and `as.data.frame.incidence()` For details, see
  https://github.com/reconhub/incidence/issues/79

incidence 1.5.0 (2018-11-01)
============================

### NEW FUNCTIONS

* `group_names()` allows the user to retrieve and set the group names.
* `get_timespan()` returns the `$timespan` element.
* `get_n()` returns the `$n` element.
* `dim()`, `nrow()`, and `ncol()` are now available for incidence objects,
  returning the dimensions of the number of bins and the number of groups.

### NEW FEATURES

* A new argument to `plot()` called `show_cases` has been added to draw borders
  around individual cases for EPIET-style curves.

R/plot.R
#' Plot function for incidence objects
#'
#' This function is used to visualise the output of the [incidence()]
#' function using the package `ggplot2`. #'
#'
#' @export
#'
#' @importFrom graphics plot
#'
#' @author Thibaut Jombart \email{thibautjombart@@gmail.com}
#'   Zhian N. Kamvar \email{zkamvar@@gmail.com}
#'
#' @seealso The [incidence()] function to generate the 'incidence'
#' objects.
#'
#' @param x An incidence object, generated by the function
#' [incidence()].
#'
#' @param fit An 'incidence_fit' object as returned by [fit()].
#'
#' @param stack A logical indicating if bars of multiple groups should be
#' stacked, or displayed side-by-side.
#'
#' @param color The color to be used for the filling of the bars; NA for
#' invisible bars; defaults to "black".
#'
#' @param border The color to be used for the borders of the bars; NA for
#' invisible borders; defaults to NA.
#'
#' @param col_pal The color palette to be used for the groups; defaults to
#' `incidence_pal1`. See [incidence_pal1()] for other palettes implemented in
#' incidence.
#'
#' @param alpha The alpha level for color transparency, with 1 being fully
#' opaque and 0 fully transparent; defaults to 0.7.
#'
#' @param xlab The label to be used for the x-axis; empty by default.
#'
#' @param ylab The label to be used for the y-axis; by default, a label will be
#' generated automatically according to the time interval used in incidence
#' computation.
#'
#' @param labels_week a logical value indicating whether labels x axis tick
#' marks are in week format YYYY-Www when plotting weekly incidence; defaults to
#' TRUE.
#'
#' @param labels_iso (deprecated) This has been superceded by `labels_iso`.
#' Previously:a logical value indicating whether labels x axis tick marks are
#' in ISO 8601 week format yyyy-Www when plotting ISO week-based weekly
#' incidence; defaults to be TRUE.
#'
#' @param show_cases if `TRUE` (default: `FALSE`), then each observation will be
#' colored by a border. The border defaults to a white border unless specified
#' otherwise. This is normally used outbreaks with a small number of cases.
#' Note: this can only be used if `stack = TRUE`
#'
#' @param n_breaks the ideal number of breaks to be used for the x-axis
#'   labeling
#' 
#' @return
#'  - `plot()` a [ggplot2::ggplot()] object.
#'  - `make_breaks()` a two-element list. The "breaks" element will contain the
#'    evenly-spaced breaks as either dates or numbers and the "labels" element
#'    will contain either a vector of weeks OR a [ggplot2::waiver()] object.
#'  - `scale_x_incidence()` a \pkg{ggplot2} "ScaleContinuous" object. 
#'
#' @details 
#'  - `plot()` will visualise an incidence object using `ggplot2`
#'  - `make_breaks()` calculates breaks from an incidence object that always
#'    align with the bins and start on the first observed incidence. 
#'  - `scale_x_incidence()` produces and appropriate `ggplot2` scale based on
#'    an incidence object. 
#' 
#' @examples
#'
#' if(require(outbreaks) && require(ggplot2)) { withAutoprint({
#'   onset <- outbreaks::ebola_sim$linelist$date_of_onset
#'
#'   ## daily incidence
#'   inc <- incidence(onset)
#'   inc
#'   plot(inc)
#'
#'   ## weekly incidence
#'   inc.week <- incidence(onset, interval = 7)
#'   inc.week
#'   plot(inc.week) # default to label x axis tick marks with isoweeks
#'   plot(inc.week, labels_week = FALSE) # label x axis tick marks with dates
#'   plot(inc.week, border = "white") # with visible border
#'
#'   ## use group information
#'   sex <- outbreaks::ebola_sim$linelist$gender
#'   inc.week.gender <- incidence(onset, interval = "1 epiweek", groups = sex)
#'   plot(inc.week.gender)
#'   plot(inc.week.gender, labels_week = FALSE)
#'
#'   ## show individual cases at the beginning of the epidemic
#'   inc.week.8 <- subset(inc.week.gender, to = "2014-06-01")
#'   p <- plot(inc.week.8, show_cases = TRUE, border = "black")
#'   p
#'
#'   ## update the range of the scale
#'   lim <- c(min(get_dates(inc.week.8)) - 7*5, 
#'            aweek::week2date("2014-W50", "Sunday"))
#'   lim
#'   p + scale_x_incidence(inc.week.gender, limits = lim)
#'
#'   ## customize plot with ggplot2
#'   plot(inc.week.8, show_cases = TRUE, border = "black") +
#'     theme_classic(base_size = 16) +
#'     theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) 
#'
#'   ## adding fit
#'   fit <- fit_optim_split(inc.week.gender)$fit
#'   plot(inc.week.gender, fit = fit)
#'   plot(inc.week.gender, fit = fit, labels_week = FALSE)
#'
#' })}
#'
plot.incidence <- function(x, ..., fit = NULL, stack = is.null(fit),
                           color = "black", border = NA, col_pal = incidence_pal1,
                           alpha = .7, xlab = "", ylab = NULL,
                           labels_week = !is.null(x$weeks),
                           labels_iso = !is.null(x$isoweeks),
                           show_cases = FALSE,
                           n_breaks = 6) {
  stopifnot(is.logical(labels_iso), is.logical(labels_week))

  the_call <- match.call()
  if (any(names(the_call) == "labels_iso"))  {
    if (any(names(the_call) == "labels_week")) {
      warning("labels_iso is deprecated. The value of `labels_week` will be used.")
    } else {
      warning("labels_iso is deprecated. Use `labels_week` instead.")
      labels_week <- labels_iso
    }
  }
  ## extract data in suitable format for ggplot2
  df <- as.data.frame(x, long = TRUE, stringsAsFactors = TRUE)
  n.groups <- ncol(x$counts)
  gnames   <- group_names(x)

  
  ## Use custom labels for usual time intervals
  if (is.null(ylab)) {
    if (is.numeric(x$interval)) {
      if (x$interval == 1) {
        ylab <- "daily incidence"
      } else if (x$interval == 7) {
        ylab <- "weekly incidence"
      } else if (x$interval == 14) {
        ylab <- "semi-weekly incidence"
      } else {
        ylab <- sprintf("incidence by period of %d days",
                        x$interval)
      }
    } else if (is.character(x$interval)) {
      
      # capturing the number and type
      p     <- "(\\d*)\\s?([a-z]+?)s?$"
      num   <- gsub(p, "\\1", tolower(x$interval))
      itype <- gsub(p, "\\2", tolower(x$interval))
      if (num == "" || num == "1") {
        ylab <- sprintf("%sly incidence", itype)
      } else {
        ylab <- sprintf("incidence by a period of %s %ss", num, itype)
      }
    }
    if (!is.null(x$weeks)) {
      type_of_week <- get_type_of_week(x)
      ylab <- gsub("(weekl?y?)", sprintf("%s \\1", type_of_week), ylab)
    }
    if (isTRUE(x$cumulative)) {
      ylab <- sub("incidence", "cumulative incidence", ylab)
    }
    first_letter          <- substring(ylab, 1, 1)
    substring(ylab, 1, 1) <- toupper(first_letter)
  }

  ## Handle stacking
  stack.txt <- if (stack) "stack" else "dodge"

  ## By default, the annotation of bars in geom_bar puts the label in the
  ## middle of the bar. This is wrong in our case as the annotation of a time
  ## interval is the lower (left) bound, and should therefore be left-aligned
  ## with the bar. Note that we cannot use position_nudge to create the
  ## x-offset as we need the 'position' argument for stacking. This can be
  ## addressed by adding interval/2 to the x-axis, but this only works until we
  ## have an interval such as "month", "quarter", or "year" where the number of
  ## days for each can vary. To alleviate this, we can create a new column that
  ## counts the number of days within each interval.

  ## Adding a variable for width in ggplot
  df$interval_days <- get_interval(x, integer = TRUE)
  ## if the date type is POSIXct, then the interval is actually interval seconds
  ## and needs to be converted to days
  if (inherits(df$dates, "POSIXct")) {
    df$interval_days <- df$interval_days * 86400 # 24h * 60m * 60s
  }
  ## Important note: it seems safest to specify the aes() as part of the geom,
  ## not in ggplot(), as it interacts badly with some other geoms like
  ## geom_ribbon - used e.g. in projections::add_projections().


  ## add mid-interval positions for x-axis

  ## THIS BREAKS THE PLOT WITH ggplot2_3.3.0
  ## See bug reports at:
  ## https://github.com/reconhub/incidence/issues/119
  ## https://github.com/tidyverse/ggplot2/issues/3873
  ## Temporary fix: changing placement to default of ggplot2::scale_x_date
  
  x_axis <- "dates + (interval_days/2)"
  y_axis <- "counts"
  out <- ggplot2::ggplot(df) +
    ggplot2::geom_col(ggplot2::aes_string(
                        x = x_axis,
                        y = y_axis
                        ),
                      width = df$interval_days,
                      position = stack.txt,
                      color = border,
                      alpha = alpha) +
    ggplot2::labs(x = xlab, y = ylab)

  ## Handle show_cases here

  if (show_cases && stack) {
    squaredf <- df[rep(seq.int(nrow(df)), df$counts), ]
    squaredf$counts <- 1
    squares <- ggplot2::geom_col(ggplot2::aes_string(
                                   x = x_axis,
                                   y = y_axis
                                   ),
                                 color = if (is.na(border)) "white" else border,
                                 fill  = NA,
                                 position = "stack",
                                 data = squaredf,
                                 width = squaredf$interval_days
                                 )
    out <- out + squares
  }
  if (show_cases && !stack) {
    message("the argument `show_cases` requires the argument `stack = TRUE`")
  }
  ## Handle fit objects here; 'fit' can be either an 'incidence_fit' object,
  ## or a list of these. In the case of a list, we add geoms one after the
  ## other.

  if (!is.null(fit)) {
    if (inherits(fit, "incidence_fit")) {
      out <- add_incidence_fit(out, fit)
    } else if (is.list(fit)) {
      for (i in seq_along(fit)) {
        fit.i <- fit[[i]]
        if (!inherits(fit.i, c("incidence_fit", "incidence_fit_list"))) {
          stop(sprintf(
            "The %d-th item in 'fit' is not an 'incidence_fit' object, but a %s",
            i, class(fit.i)))
        }
        out <- add_incidence_fit(out, fit.i)
      }
    } else {
      stop("Fit must be a 'incidence_fit' object, or a list of these")
    }
  }


  ## Handle colors

  ## Note 1: because of the way 'fill' works, we need to specify it through
  ## 'aes' if not directly in the geom. This causes the kludge below, where we
  ## make a fake constant group to specify the color and remove the legend.

  ## Note 2: when there are groups, and the 'color' argument does not have one
  ## value per group, we generate colors from a color palette. This means that
  ## by default, the palette is used, but the user can manually specify the
  ## colors.

  if (n.groups < 2 && is.null(gnames)) {
    out <- out + ggplot2::aes(fill = 'a') +
      ggplot2::scale_fill_manual(values = color, guide = FALSE)
  } else {
    if (!is.null(names(color))) {
      tmp     <- color[gnames] 
      matched <- names(color) %in% names(tmp)
      if (!all(matched)) {
        removed <- paste(names(color)[!matched], 
                         color[!matched],
                         sep = '" = "',
                         collapse = '", "')
        message(sprintf("%d colors were not used: \"%s\"", sum(!matched), removed))
      }
      color <- tmp
    }
                                 
    ## find group colors
    if (length(color) != n.groups) {
      msg <- "The number of colors (%d) did not match the number of groups (%d)"
      msg <- paste0(msg, ".\nUsing `col_pal` instead.")
      default_color <- length(color) == 1L && color == "black"
      if (!default_color) {
        message(sprintf(msg, length(color), n.groups))
      }
      group.colors <- col_pal(n.groups)
    } else {
      group.colors <- color
    }

    ## add colors to the plot
    out <- out + ggplot2::aes_string(fill = "groups") +
      ggplot2::scale_fill_manual(values = group.colors)
    if (!is.null(fit)) {
      out <- out + ggplot2::aes_string(color = "groups") +
        ggplot2::scale_color_manual(values = group.colors)
    }
  }

  out <- out + scale_x_incidence(x, n_breaks, labels_week, ...)
  out
}


## This function will take an existing 'incidence' plot object ('p') and add lines from an
## 'incidence_fit' object ('x')

#' @export
#' @rdname plot.incidence
#'
#' @param p An existing incidence plot.
add_incidence_fit <- function(p, x, col_pal = incidence_pal1){
  if (inherits(x, "incidence_fit_list")) {
    x <- get_fit(x)
  }
  ## 'x' could be a list of fit, in which case all fits are added to the plot
  if (is.list(x) && !inherits(x, "incidence_fit")) {
    out <- p
    for (e in x) {
      if (inherits(e, "incidence_fit")) {
        out <- add_incidence_fit(out, e, col_pal)
      }
    }
    return(out)
  }
  df <- get_info(x, "pred")

  # In the case that the incidence object is of the type POSIXt, the data from
  # the fit object must be presented as POSIXt or ggplot2 will throw a fit.

  if (inherits(p$data$dates, "POSIXt")) {
    # I add half a day here because any fractional days (0.5) will be thrown out
    # on conversion and I'm not quite sure why that is
    df$dates <- as.POSIXlt(df$dates) + 43200 # adding half a day
    if (inherits(p$data$dates, "POSIXct")) {
      df$dates <- as.POSIXct(df$dates)
    } 
  }

  out <- suppressMessages(
    p + ggplot2::geom_line(
      data = df,
      ggplot2::aes_string(x = "dates", y = "fit"), linetype = 1) +
      ggplot2::geom_line(
        data = df,
        ggplot2::aes_string(x = "dates", y = "lwr"), linetype = 2) +
      ggplot2::geom_line(
        data = df,
        ggplot2::aes_string(x = "dates", y = "upr"), linetype = 2)
  )


  if ("groups" %in% names(df)) {
    n.groups <- length(levels(df$groups))
    out <- out + ggplot2::aes_string(color = "groups") +
      ggplot2::scale_color_manual(values = col_pal(n.groups))
  }

  out
}





#' @export
#' @rdname plot.incidence

plot.incidence_fit <- function(x, ...){
  base <- ggplot2::ggplot()
  out <- add_incidence_fit(base, x, ...) +
    ggplot2::labs(x = "", y = "Predicted incidence")
  out
}

#' @export
#' @rdname plot.incidence

plot.incidence_fit_list <- function(x, ...){
  base <- ggplot2::ggplot()
  fits <- get_fit(x)
  out  <- add_incidence_fit(base, fits, ...) +
    ggplot2::labs(x = "", y = "Predicted incidence")
  out
}

